File: libs/blueprint/src/main/java/instructure/androidblueprint/PresenterFragment.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package instructure.androidblueprint;


import android.os.Bundle;
import com.instructure.canvasapi2.models.Section;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

public abstract class PresenterFragment<
        PRESENTER extends FragmentPresenter<VIEW>,
        VIEW extends FragmentViewInterface> extends Fragment {

    protected abstract void onReadySetGo(PRESENTER presenter);

    protected abstract PresenterFactory<PRESENTER> getPresenterFactory();

    protected abstract void onPresenterPrepared(PRESENTER presenter);

    private static final int LOADER_ID = 1003;

    private boolean mDelivered = false;
    private Presenter<VIEW> mPresenter;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        getLoaderManager().initLoader(LOADER_ID, null, new LoaderManager.LoaderCallbacks<PRESENTER>() {
            @Override
            public Loader<PRESENTER> onCreateLoader(int id, Bundle args) {
                return new PresenterLoader<>(getContext(), getPresenterFactory());
            }

            @Override
            public void onLoadFinished(Loader<PRESENTER> loader, PRESENTER presenter) {
                if (!mDelivered) {
                    PresenterFragment.this.mPresenter = presenter;
                    mDelivered = true;
                    onPresenterPrepared(presenter);
                }
            }

            @Override
            public void onLoaderReset(Loader<PRESENTER> loader) {
                PresenterFragment.this.mPresenter = null;
                onPresenterDestroyed();
            }
        });
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onResume() {
        super.onResume();
        onReadySetGo((PRESENTER) mPresenter.onViewAttached(getPresenterView()));
    }

    @Override
    public void onPause() {
        mPresenter.onViewDetached();
        super.onPause();
    }

    protected void onPresenterDestroyed() {
        // hook for subclasses
    }

    // Override in case of fragment no implementing Presenter<VIEW> interface
    @SuppressWarnings("unchecked")
    protected VIEW getPresenterView() {
        return (VIEW) this;
    }

    @SuppressWarnings("unchecked")
    protected PRESENTER getPresenter() {
        return (PRESENTER) mPresenter;
    }

    protected void addSwipeToRefresh(@NonNull SwipeRefreshLayout swipeRefreshLayout) {
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getPresenter().refresh(true);
            }
        });
    }
}

File: libs/pandautils/src/main/java/com/instructure/pandautils/blueprint/PresenterFragment.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */
package com.instructure.pandautils.blueprint

import android.os.Bundle
import com.instructure.pandautils.base.BaseCanvasFragment
import androidx.loader.app.LoaderManager
import androidx.loader.content.Loader
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

abstract class PresenterFragment<PRESENTER : FragmentPresenter<VIEW>, VIEW : FragmentViewInterface> : BaseCanvasFragment() {

    protected abstract fun onReadySetGo(presenter: PRESENTER)

    protected abstract fun getPresenterFactory(): PresenterFactory<VIEW, PRESENTER>

    protected abstract fun onPresenterPrepared(presenter: PRESENTER)

    private var delivered = false

    lateinit var presenter: PRESENTER
        private set

    @Suppress("DEPRECATION")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loaderManager.initLoader(
            LOADER_ID,
            null,
            object : LoaderManager.LoaderCallbacks<PRESENTER> {
                override fun onCreateLoader(id: Int, args: Bundle?): Loader<PRESENTER> {
                    return PresenterLoader(requireContext(), getPresenterFactory())
                }

                override fun onLoadFinished(
                    loader: Loader<PRESENTER>,
                    presenter: PRESENTER
                ) {
                    if (!delivered) {
                        this@PresenterFragment.presenter = presenter
                        delivered = true
                        onPresenterPrepared(presenter)
                    }
                }

                override fun onLoaderReset(loader: Loader<PRESENTER>) {
                    //presenter = null
                    onPresenterDestroyed()
                }
            })
    }


    @Suppress("UNCHECKED_CAST")
    override fun onResume() {
        super.onResume()
        onReadySetGo(presenter.onViewAttached(presenterView) as PRESENTER)
    }

    override fun onPause() {
        presenter.onViewDetached()
        super.onPause()
    }

    protected fun onPresenterDestroyed() {
        // hook for subclasses
    }

    // Override in case of fragment no implementing Presenter<VIEW> interface
    @Suppress("UNCHECKED_CAST")
    private val presenterView: VIEW
        get() = this as VIEW


    protected fun addSwipeToRefresh(swipeRefreshLayout: SwipeRefreshLayout) {
        swipeRefreshLayout.setOnRefreshListener { presenter.refresh(true) }
    }

    companion object {
        private const val LOADER_ID = 1003
    }
}
